src/org/labkey/test/tests/InvalidateSessionTest.java

```diff
@@ -0,0 +1,151 @@
+    {
+        signOut();
+        signIn(USER);
+        Connection cn = createDefaultConnection();
+        SelectRowsResponse response;
+        SelectRowsCommand selectCmd = new SelectRowsCommand("auditLog", "UserAuditEvent");
+        try
+        {
+            response = selectCmd.execute(cn, "Home");
+            Assert.assertEquals("Did not establish the database connection before the password change", 200,
+                    response.getStatusCode());
+        }
+        catch (IOException | CommandException e)
+        {
+            throw new RuntimeException(e);
+        }
+
+        log("Changing the user password");
+        String newPassword = PasswordUtil.getPassword() + "&*&*";
+        goToMyAccount().clickChangePassword()
+                .setOldPassword(PasswordUtil.getPassword())
+                .setPassword1(newPassword)
+                .setPassword2(newPassword)
+                .clickSubmit();
+
+        log("Verifying the session is invalid");
+        try
+        {
+            selectCmd.execute(cn, getProjectName());
+            fail("Test should have thrown command exception");
+        }
+        catch (CommandException e)
+        {
+            Assert.assertEquals("Session should be expired", 401, e.getStatusCode());
+        }
+        signOut();
+    }
+
+    /*
+        Regression coverage for Secure Issue 31493: Test for session and cookie persistence through login and logout
+     */
+    @Test
+    public void testCookieAndSessionFromLogout() throws IOException
+    {
+        log("Capture the cookie after login");
+        Cookie beforeCookie = getCookies(getCurrentUser()).get(Connection.JSESSIONID);
+
+        log("Establish the connection");
+        Connection cn = createDefaultConnection();
+        SelectRowsResponse response;
+        SelectRowsCommand selectCmd = new SelectRowsCommand("auditLog", "UserAuditEvent");
+        try
+        {
+            response = selectCmd.execute(cn, getProjectName());
+            Assert.assertEquals("Did not establish the database connection before the password change", 200,
+                    response.getStatusCode());
+        }
+        catch (CommandException e)
+        {
+            throw new RuntimeException(e);
+        }
+
+        log("Sign out");
+        signOut();
+
+        log("Verify previously established connection be used after log out");
+        try
+        {
+            selectCmd.execute(cn, getProjectName());
+            fail("Should fail when logged out");
+        }
+        catch (CommandException e)
+        {
+            Assert.assertEquals("Command execution should fail when logged out", HttpStatus.SC_UNAUTHORIZED, e.getStatusCode());
+        }
+
+        log("Capture the cookie after logout");
+        Cookie afterCookie = getCookies(getCurrentUser()).get(Connection.JSESSIONID);
+        Assert.assertFalse("Before and after log out cookie should be different", beforeCookie.equals(afterCookie));
```

Comment:
Sorry, I should have been more specific. Just verify the `value` of the cookie instead of the entire cookie object. I wouldn't count on `Cookie.equals` to do what this test wants. For example, the expiration date of the cookie might change without the value changing.